src/main/java/com/amazon/ion/impl/bin/IonRawBinaryWriter_1_1.kt

```diff
@@ -114,50 +140,155 @@ class IonRawBinaryWriter_1_1 internal constructor(
      */
     private inline fun openValue(valueWriterExpression: () -> Unit) {
-        if (numAnnotations > 0) {
-            TODO("Actually write out the annotations.")
+
+        // Start at 1, assuming there's an annotations OpCode byte.
+        // We'll clear this if there are no annotations.
+        var annotationsTotalLength = 1L
+
+        // Effect of the xor: if annotationsFlexSymFlag is -1, then we're matching `-1 * numAnnotations - 1`
+        when (numAnnotations xor annotationFlexSymFlag) {
+            0, -1 -> annotationsTotalLength = 0
+            1 -> {
+                buffer.writeByte(OpCodes.ANNOTATIONS_1_SYMBOL_ADDRESS)
+                annotationsTotalLength += buffer.writeFlexUInt(annotationsIdBuffer[0])
+            }
+            2 -> {
+                buffer.writeByte(OpCodes.ANNOTATIONS_2_SYMBOL_ADDRESS)
+                annotationsTotalLength += buffer.writeFlexUInt(annotationsIdBuffer[0])
+                annotationsTotalLength += buffer.writeFlexUInt(annotationsIdBuffer[1])
+            }
+            -2 -> {
+                // If there's only one annotation, and we know that at least one has text, we don't need to check
+                // whether this is SID.
+                buffer.writeByte(OpCodes.ANNOTATIONS_1_FLEX_SYM)
+                annotationsTotalLength += buffer.writeFlexSym(utf8StringEncoder.encode(annotationsTextBuffer[0].toString()))
+                annotationsTextBuffer[0] = null
+            }
+            -3 -> {
+                buffer.writeByte(OpCodes.ANNOTATIONS_2_FLEX_SYM)
+                annotationsTotalLength += writeFlexSymFromAnnotationsBuffer(0)
+                annotationsTotalLength += writeFlexSymFromAnnotationsBuffer(1)
+            }
+            else -> annotationsTotalLength += writeManyAnnotations()
         }
+        currentContainer.length += annotationsTotalLength
+
         numAnnotations = 0
+        annotationFlexSymFlag = 0
         hasFieldName = false
         valueWriterExpression()
     }
 
+    /**
+     * Writes a FlexSym annotation for the specified position in the annotations buffers.
+     */
+    private fun writeFlexSymFromAnnotationsBuffer(i: Int): Int {
+        val annotationText = annotationsTextBuffer[i]
+        return if (annotationText != null) {
+            annotationsTextBuffer[i] = null
+            buffer.writeFlexSym(utf8StringEncoder.encode(annotationText.toString()))
+        } else {
+            buffer.writeFlexSym(annotationsIdBuffer[i])
+        }
+    }
+
+    /**
+     * Writes 3 or more annotations for SIDs or FlexSyms
+     */
+    private fun writeManyAnnotations(): Long {
+        currentContainer = containerStack.push {
+            it.clear()
+            it.type = Annotations
+            it.position = buffer.position()
+        }
+        if (annotationFlexSymFlag == FLEX_SYMS_REQUIRED) {
+            buffer.writeByte(OpCodes.ANNOTATIONS_MANY_FLEX_SYM)
+            buffer.reserve(lengthPrefixPreallocation)
```

Comment:
I wonder if this should be capped at 1 rather than using the same value used for containers. It seems less likely that your annotations would grow larger than can be described in a 1-byte FlexUInt.

That said, making a change here is probably only valuable if it simplifies the code.

Reply:
It does marginally simplify the code because the minimum we always need to preallocate is 1, so if we're also capping it at 1, then we can just hard-code it to 1.